subprojects/gradle/runner/service/src/main/kotlin/com/avito/runner/service/worker/DeviceWorker.kt

```diff
@@ -33,53 +33,60 @@ class DeviceWorker(
-
                 device.log("Receive intention: $intention")
-
-                device.log("Preparing state: ${intention.state}")
-                state = prepareDeviceState(currentState = state, intentionState = intention.state).get()
-                device.log("State prepared")
-
-                val result = executeAction(action = intention.action)
-                device.log("Worker test run completed for intention")
-                messagesChannel.send(
-                    DeviceWorkerMessage.Result(
-                        intentionResult = IntentionResult(
-                            intention = intention,
-                            actionResult = InstrumentationTestRunActionResult(
-                                testCaseRun = result
+                when (val status = device.deviceStatus()) {
+                    is Device.DeviceStatus.Alive -> {
+                        device.log("Preparing state: ${intention.state}")
+                        state = prepareDeviceState(currentState = state, intentionState = intention.state).get()
+                        device.log("State prepared")
+
+                        val result = executeAction(action = intention.action)
+                        device.log("Worker test run completed for intention")
+                        messagesChannel.send(
+                            DeviceWorkerMessage.Result(
+                                intentionResult = IntentionResult(
+                                    intention = intention,
+                                    actionResult = InstrumentationTestRunActionResult(
+                                        testCaseRun = result
+                                    )
+                                )
+                            )
+                        )
+                    }
+                    is Device.DeviceStatus.Freeze -> {
+                        // means device worker is die. TODO release device
+                        device.warn("Device died. Can't process intention: $intention", status.reason)
+
+                        messagesChannel.send(
+                            DeviceWorkerMessage.WorkerFailed(
+                                t = status.reason,
+                                intention = intention
                             )
                         )
-                    )
-                )
 
+                        return@launch
+                    }
+                }
             } catch (t: Throwable) {
-                // means device worker is die. TODO release device
-                device.warn("Error during $intention processing", t)
-
-                messagesChannel.send(
-                    DeviceWorkerMessage.WorkerFailed(
-                        t = t,
-                        intention = intention
-                    )
-                )
-
-                return@launch
+                throw RuntimeException("Unexpected error while process intention: $intention", t)
```

Comment:
It will kill the whole InstrumentationTask. After that whole deployment will be released
Now there is no situation that can throw an exception from `executeAction ` or `device.deviceStatus()`

Reply:
I don't understand this decision.
If one DeviceWorker is failed, why do we fail the whole task?

Any interaction with ADB might fail. E.g. `prepareDeviceState`